types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,233 @@
+		opt(snp)
+	}
+
+	return snp
+}
+
+// SenderNonceSeedOpt Option To add a Seed for random type when calling the constructor NewSenderNonceMempool
+// Example:
+// > random_seed := int64(1000)
+// > NewSenderNonceMempool(SenderNonceSeedTxOpt(random_seed))
+func SenderNonceSeedOpt(seed int64) SenderNonceOptions {
+	return func(snp *senderNonceMempool) {
+		snp.setSeed(seed)
+	}
+}
+
+// SenderNonceMaxTxOpt Option To set limit of max tx when calling the constructor NewSenderNonceMempool
+// Example:
+// > NewSenderNonceMempool(SenderNonceMaxTxOpt(100))
+func SenderNonceMaxTxOpt(maxTx int) SenderNonceOptions {
+	return func(snp *senderNonceMempool) {
+		snp.maxTx = maxTx
+	}
+}
+
+func (snm *senderNonceMempool) setSeed(seed int64) {
+	s1 := rand.NewSource(seed)
+	snm.rnd = rand.New(s1) //#nosec // math/rand is seeded from crypto/rand by default
+}
+
+// Insert adds a tx to the mempool. It returns an error if the tx does not have at least one signer.
+// priority is ignored.
+func (snm *senderNonceMempool) Insert(_ sdk.Context, tx sdk.Tx) error {
+	if snm.maxTx > 0 && snm.CountTx() >= snm.maxTx {
+		return ErrMempoolTxMaxCapacity
+	}
+	if snm.maxTx < 0 {
+		return nil
+	}
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	senderTxs, found := snm.senders[sender]
+	if !found {
+		senderTxs = skiplist.New(skiplist.Uint64)
+		snm.senders[sender] = senderTxs
+	}
+	senderTxs.Set(nonce, tx)
+	key := txKey{nonce: nonce, address: sender}
+	snm.existingTx[key] = true
+	return nil
+}
+
+// Select returns an iterator ordering transactions the mempool with the lowest nonce of a random selected sender first.
+func (snm *senderNonceMempool) Select(_ sdk.Context, _ [][]byte) Iterator {
+	var senders []string
+	senderCursors := make(map[string]*skiplist.Element)
+
+	orderedSenders := skiplist.New(skiplist.String)
+
+	// #nosec
+	for s := range snm.senders {
+		orderedSenders.Set(s, s)
+	}
+
+	s := orderedSenders.Front()
+	for s != nil {
+		sender := s.Value.(string)
+		senders = append(senders, sender)
+		senderCursors[sender] = snm.senders[sender].Front()
+		s = s.Next()
+	}
+
+	iter := &senderNonceMepoolIterator{
+		senders:       senders,
+		rnd:           snm.rnd,
+		senderCursors: senderCursors,
+	}
+
+	return iter.Next()
+}
+
+// CountTx returns the total count of txs in the mempool.
+func (snm *senderNonceMempool) CountTx() int {
+	return len(snm.existingTx)
+}
+
+// Remove removes a tx from the mempool. It returns an error if the tx does not have at least one signer or the tx
+// was not found in the pool.
+func (snm *senderNonceMempool) Remove(tx sdk.Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
```

Comment:
Why are we checking for this in the removal? If a tx somehow got into the mempool with no signers, we should want to be able to remove that, right?

Reply:
We need to get the sender in order to properly remove the tx. If we cannot get the sender, we cannot remove.